pact/kpenny/kpenny.pact

```diff
@@ -0,0 +1,241 @@
+      receiver:string
+      amount:decimal
+    )
+    @managed amount TRANSFER-mgr
+    (enforce-valid-transfer sender receiver (precision) amount)
+    (compose-capability (DEBIT sender))
+    (compose-capability (CREDIT receiver))
+  )
+
+  (defun TRANSFER-mgr:decimal
+    ( managed:decimal
+      requested:decimal
+    )
+
+    (let ((newbal (- managed requested)))
+      (enforce (>= newbal 0.0)
+        (format "TRANSFER exceeded for balance {}" [managed]))
+      newbal)
+  )
+
+  (defconst MINIMUM_PRECISION:integer 14)
+
+  (defconst CONVERSION_RATE:decimal 1000000.0)
+
+  (defconst KPENNY_BANK:string 'kpenny-bank)
+
+  (defconst SWAP_DEADLINE (time "2021-03-15T00:00:00Z"))
+
+  (defconst FINAL_DEADLINE (time "2021-03-30T00:00:00Z"))
+
+  (defun kpenny-bank-guard () (create-module-guard 'kpenny-admin))
+
+  (defun init ()
+    (coin.create-account KPENNY_BANK (kpenny-bank-guard))
+  )
+
+  (defun enforce-unit:bool (amount:decimal)
+    (enforce-precision (precision) amount))
+
+  (defun create-account:string
+    ( account:string
+      guard:guard
+    )
+    (enforce-valid-account account)
+    (insert ledger account
+      { "balance" : 0.0
+      , "guard"   : guard
+      })
+    )
+
+  (defun get-balance:decimal (account:string)
+    (at 'balance (read ledger account))
+  )
+
+  (defun details:object{fungible-v2.account-details}
+    ( account:string )
+    (with-read ledger account
+      { "balance" := bal
+      , "guard" := g }
+      { "account" : account
+      , "balance" : bal
+      , "guard": g })
+    )
+
+  (defun rotate:string (account:string new-guard:guard)
+    (with-read ledger account
+      { "guard" := old-guard }
+
+      (enforce-guard old-guard)
+
+      (update ledger account
+        { "guard" : new-guard }))
+    )
+
+
+  (defun fund:string (account:string amount:decimal guard:guard)
+    (with-capability (CREDIT account)
+      (require-capability (FUND))
+        (credit account guard amount))
+  )
+
+  (defun convert:string (account:string amount-kda:decimal)
+    (enforce-guard (before-date SWAP_DEADLINE))
+    (coin.transfer account KPENNY_BANK amount-kda)
+    (let
+      ( (tx-id (hash {"account": account, "amount": amount-kda, "salt": (at "block-time" (chain-data))}))
+        (amount-kpenny (* amount-kda CONVERSION_RATE))
+        (g (at 'guard (coin.details account)))
+      )
+      (insert conversions (format "{}-{}" [account, tx-id])
+        { "account"        : account
+        , "amount-kda"     : amount-kda
+        , "amount-kpenny"  : amount-kpenny
+        })
+      (with-capability (FUND)
+        (fund account amount-kpenny g))
+    )
+  )
+
+  (defun redeem-one:string (account:string)
+    (require-capability (REDEEM))
+    (with-read ledger account
+      { "balance" := amount-kpenny,
+        "guard" := g
+      }
+      (if (= amount-kpenny 0.0)
+        "Account's kpenny balance is 0.0"
+        (coin.transfer-create KPENNY_BANK account g (/ amount-kpenny CONVERSION_RATE))
```

Comment:
you'll need to check (and test in repl) KDA precision enforcement

Reply:
We shouldn't do `transfer-create` according to current semantics of reservations, where you have the KDA account already, and the KDA account name is your KPenny name. By this logic we should only use `transfer` here

Reply:
After discussion, I got rid of `redeem-all`/`redeem-one` and instead one `redeem` function that only the user can call to any account with provided guard.